code/modules/research/xenobiology/crossbreeding/warping.dm

```diff
@@ -0,0 +1,1227 @@
+	///current x,y,z location of the reserved space for the rune room
+	var/datum/turf_reservation/room_reservation
+	///the template of the warped_room map
+	var/datum/map_template/warped_room/rune_room
+	///list of people that teleported into the rune_room. The room will dissapear if the list is empty and the rune is destroyed.
+	var/list/customer_list
+
+
+/obj/effect/warped_room_exit
+	name = "warped_rune"
+	icon = 'icons/obj/slimecrossing.dmi'
+	icon_state = "rainbow_rune"
+	desc = "Use this rune if you want to leave this place. You will have to leave eventually."
+	move_resist = INFINITY
+	anchored = TRUE
+	///where the rune will teleport you back.
+	var/turf/exit_turf
+	///rune linked to the exit rune
+	var/obj/effect/warped_rune/rainbowspace/enter_rune
+
+
+/datum/map_template/warped_room
+	name = "Warped room"
+	mappath = '_maps/templates/warped_room.dmm'
+
+
+/area/warped_room
+	name = "warped room"
+	icon_state = "yellow"
+	dynamic_lighting = DYNAMIC_LIGHTING_FORCED
+	requires_power = FALSE
+	has_gravity = TRUE
+	noteleport = TRUE
+
+
+///creates the warped room and place an exit rune to exit the room
+/obj/effect/warped_rune/rainbowspace/Initialize()
+	. = ..()
+	rune_room = new()
+	room_reservation = SSmapping.RequestBlockReservation(rune_room.width, rune_room.height) //monkey sees valid location
+	rune_room.load(locate(room_reservation.bottom_left_coords[1], room_reservation.bottom_left_coords[2], room_reservation.bottom_left_coords[3]))//monkey room activate
+	var/obj/effect/warped_room_exit/exit_rune = new(locate(room_reservation.bottom_left_coords[1] + 3, room_reservation.bottom_left_coords[2] + 6, room_reservation.bottom_left_coords[3]))
+	exit_rune.exit_turf = rune_turf //we spawn the exit rune manually in the code because it's easier to link it to the rainbow rune that way.
+	exit_rune.enter_rune = src
+
+
+///here to check if anyone's being transported in or out of the room with the user.
+/obj/effect/warped_rune/rainbowspace/proc/customer_check(atom/person_checked ,smuggle_in)
+	var/list/hidden_customers = person_checked.GetAllContents(/mob/living/carbon/human)
+	if(!length(hidden_customers))
+		return
+	for(var/mob/living/carbon/human/customer in hidden_customers)
+		if(smuggle_in)
+			LAZYADD(customer_list, customer) //if they enter the room
+		else
+			LAZYREMOVE(customer_list, customer) //if they exit the room
+
+
+/obj/effect/warped_rune/rainbowspace/attack_hand(mob/living/user)
+	. = ..()
+	for(var/mob/living/carbon/human/customer in rune_turf)
+		customer.forceMove(locate(room_reservation.bottom_left_coords[1] + 3, room_reservation.bottom_left_coords[2] + 6, room_reservation.bottom_left_coords[3]))
+		customer_check(customer, TRUE)
+
+
+///Will delete the room when the rune is destroyed if no customer is left in the room.
+/obj/effect/warped_rune/rainbowspace/Destroy()
+	if(!length(customer_list))
+		qdel(room_reservation)
+	return ..()
```

Comment:
Clean the references to `customer_list`, `room_reservation` and `rune_room` on deletion.
If they can be destroyed before the rune, then register to their destruction and clean them up as well.

Reply:
Also, since this is a lazy list, use `LAZYLEN` instead of `length`
It's quite the same, but it indicates it being one.